Fix issues #4 and #5. Also add another common MBIM port value for DEV (as a comment).

AndySchroder opened the pull request:
Fix issues #4 and #5. Also add another common MBIM port value for DEV (as a comment).

AndySchroder committed:
fix issue #4

AndySchroder committed:
add another common MBIM port value for DEV (as a comment)

AndySchroder committed:
add installation instructions for mbimcli on ubuntu

stich86 reviewed (changes requested):
I think it's better to put in the comments instead of create new variable like:

`# Point to MBIM port (ex. /dev/cdc-wdmX or /dev/wwan0mbim0`

stich86 reviewed (changes requested):
Are you sure about this one?

the channel_id is passed as parameter in function close_channel, `param` is the ID got from open_channel

```
def open_channel(aid: str) -> str:
    output_cmd = run_mbimcli(f'--ms-set-uicc-open-channel="application-id={aid}"').split('\n')
    channel_line = next(line for line in output_cmd if line.startswith('\t channel:'))
    channel_id_str = channel_line.split(': ')[1].strip()
    channel_id = int(channel_id_str)
    return channel_id

def close_channel(channel_id: int) -> None:
    run_mbimcli(f'--ms-set-uicc-close-channel="channel={channel_id}"')
```

I'll re-test it in the next days, but i'm sure it's working as expected because on my case, if i'm not closing channel correctly I got each time error

AndySchroder commented:
The reason I did it this way was because if using the `vi` editor, you can quickly switch between the two options by doing `dd` to delete the line of the active variable value and `x` to uncomment inactive variable value. Or, you can just leave the first line there because uncommenting the second variable value just overwrites the first when running.

AndySchroder commented:
At https://github.com/stich86/lpac-libmbim-wrapper/blob/7c282a675472d973a7d06a961b9dcdd69fe3f634/wrapper.py#L17 it indicates the channel_id is never provided again. The `param` you get seems to be something different. I think that variable name is used in about 3 different ways in the script, I think that is why the bug was created.

stich86 commented:
> At
> 
> https://github.com/stich86/lpac-libmbim-wrapper/blob/7c282a675472d973a7d06a961b9dcdd69fe3f634/wrapper.py#L17
> 
> it indicates the channel_id is never provided again. The `param` you get seems to be something different. I think that variable name is used in about 3 different ways in the script, I think that is why the bug was created.

i'll try in the next days to check and i let you know ;)

AndySchroder committed:
renamed wrapper.py to lpac-mbim

AndySchroder committed:
moved device definition to the command line, created setup.cfg, and updated install instructions

AndySchroder committed:
fixed typo

AndySchroder commented:
Also added 78346b47400801938a8dc8c79f525ecff2178400 , "moved device definition to the command line, created setup.cfg, and updated install instructions". I think this should make the script a lot easier to use.

AndySchroder committed:
fixed a find and replace mistake in README.md

AndySchroder committed:
remove install_requires from setup.cfg because all modules listed were actually standard modules and not available on PyPI

AndySchroder committed:
fixed command line parser

AndySchroder committed:
recommend installing and running as root because the normal user likely will not have access to the device port

AndySchroder committed:
Merge branch 'main' of https://github.com/stich86/lpac-libmbim-wrapper

bam80 commented:
@AndySchroder I see you got involved into this project, would you like to try to solve #7?

AndySchroder commented:
> @AndySchroder I see you got involved into this project, would you like to try to solve #7?

I'm happy to test it, but I don't have the capability to do more than that. The changes I propose in this PR make a clean and workable enough solution for me right now. #7 will be great to further simplify of course, so I'll be glad when this wrapper becomes obsolete, but the wrapper does work well.

bam80 commented:
Any update on this PR?

stich86 commented:
I should receive next week some LTE modules, so I can test the PR on it

AndySchroder commented:
I have no more changes planned. Would appreciate it being merged. If no merge, I'll continue my fork and change the reference download path for the pip command to point to my fork.

AndySchroder commented:
What model module will you be testing with?

stich86 commented:
> I have no more changes planned. Would appreciate it being merged. If no merge, I'll continue my fork and change the reference download path for the pip command to point to my fork.

This is a free time project, and as I've said want to test it with these new modules (Quectel EP-06, a 4G Sierra Wireless and a Huawei module)

If you're in a hurry, go ahead with your fork, nothing prevents you from doing so :)

stich86 commented:
merged, should receive next week some modules
So I'll update the table ;)

AndySchroder commented:
so you were able to test with an existing module you already have, or you didn't test at all yet?

stich86 commented:
> so you were able to test with an existing module you already have, or you didn't test at all yet?

not yet..
my friend had a problem.. hope he will send me modules by end of this week :(

z3ntu commented on lpac-mbim:
```diff
@@ -76,8 +83,8 @@ def handle_type_apdu(func: str, param: str):
-            print(f"INFO: Close channel {param}")
-            close_channel(int(param))
+            print(f"INFO: Close channel {CHANNEL_ID}")
+            close_channel(int(CHANNEL_ID))
```
I think this is not particularly correct (well it works), but lpac is returning a bogus channel number because you're not giving it the correct channel number in logic_channel_open. You're always returning 1 there while the return value should be the channel number it seems https://github.com/z3ntu/lpac-libqmi-wrapper/blob/main/lpac-libqmi-wrapper.py#L90

stich86 replied:
The original code had your suggestion, but @AndySchroder said that for him was not working. Very strange because I tested it and never had a problem.

I'll check it again when receive new modules, in case i'll revert this change back to your original one

AndySchroder replied:
I think the old way was always returning 1, but the new way I have changed it to uses the correct channel id.

AndySchroder replied:
I don't think you can specify the channel number in logic_channel_open. It's not that it is not specified correctly, it isn't an input at all, only an output. Is that correct?

AndySchroder replied:
As mentioned in https://github.com/stich86/lpac-libmbim-wrapper/pull/6#issuecomment-2132291278 , the `param` variable is used in about 3 different ways, I think that is why this bug was created.

AndySchroder replied:
The application id is the input to logic_channel_open, not the channel id.

AndySchroder replied:
Wait, you are saying that

https://github.com/stich86/lpac-libmbim-wrapper/blob/a4a2c7b4b7cf2f01f981608af1f5141895ec5084/lpac-mbim#L79

is wrong and should match

https://github.com/z3ntu/lpac-libqmi-wrapper/blob/main/lpac-libqmi-wrapper.py#L90

?

If that change is made, then lpac will keep track of the channel id correctly and pass it around in `param` instead of the wrapper needed to do it itself? If so, I did not realize that lpac did that and so that is why I figured the `param` variable was not being used properly in all places.

z3ntu replied:
> Wait, you are saying that
> 
> https://github.com/stich86/lpac-libmbim-wrapper/blob/a4a2c7b4b7cf2f01f981608af1f5141895ec5084/lpac-mbim#L79
> 
> is wrong and should match
> 
> https://github.com/z3ntu/lpac-libqmi-wrapper/blob/main/lpac-libqmi-wrapper.py#L90
> 
> ?
> 
> If that change is made, then lpac will keep track of the channel id correctly and pass it around in `param` instead of the wrapper needed to do it itself? If so, I did not realize that lpac did that and so that is why I figured the `param` variable was not being used properly in all places.

Yes, I'm saying that. Try it yourself, return `"ecode": 2`, and you should then get this parameter in the close function

I didn't really check the code inside lpac itself, but from my observation the return value from channel_open will be the parameter you get for channel_close.

AndySchroder replied:
okay, but why do you use the global variable `CHANNEL_ID` in the `send_apdu` function then?

AndySchroder replied:
It seems as though lpac will only pass the channel id as the parameter for closing the channel, but nothing else?

z3ntu replied:
Because there we don't get the channel id again, so we need to persist it ourselves regardless. Imo the channel id should also be provided by lpac as parameter for that function, but they don't ..

AndySchroder replied:
okay, so is the fix I've done more consistent with the rest of the logic in the wrapper required or should we change it?
